red53_agent.go

```diff
@@ -0,0 +1,203 @@
+			Changes: []*route53.Change{
+				{
+					Action: aws.String(route53.ChangeActionUpsert),
+					ResourceRecordSet: &route53.ResourceRecordSet{
+						Name: aws.String(serviceName + "." + DNSName),
+						// It creates an A record with the name of the service
+						Type:            aws.String(route53.RRTypeA),
+						ResourceRecords: aValues,
+						TTL:             aws.Int64(defaultTTL),
+					},
+				},
+			},
+			Comment: aws.String("Service Discovery Created Record"),
+		},
+		HostedZoneId: aws.String(configuration.HostedZoneId),
+	}
+	_, err = r53.ChangeResourceRecordSets(params)
+	logErrorNoFatal(err)
+	log.Debug("Record " + serviceName + "." + DNSName + " updated.")
+	return err
+}
+
+var redisClient *redis.Client
+
+func processServicePings(servicePings []string) {
+	var serviceMap = make(map[string][]string)
+
+	for _, servicePing := range servicePings {
+		pingChunks := strings.Split(servicePing, "_")
+		serviceName := pingChunks[0]
+		serviceIp := pingChunks[1]
+		serviceMap[serviceName] = append(serviceMap[serviceName], serviceIp)
+	}
+
+	for serviceName, serviceIps := range serviceMap {
+		modifyDNSRecord(serviceName, serviceIps)
+	}
+}
+
+func fetchActiveServices() {
+	var servicePings []string
+	var sum int
+	var err error
+	log.Debug("Fetching active services ...")
+
+	// Fetch last "fetchLast" seconds of service pings
+	now := time.Now().Unix()
+	epoch := now - fetchLast // seconds
+
+	// Try several times with exponential backoff.
+	sum = 1
+	for {
+		servicePings, err = redisClient.ZRange("redecs:service_pings", epoch, now).Result()
+
+		if err == nil {
+			break
+		}
+		if sum > 8 {
+			// Bail out if this is failing.
+			logErrorAndFail(err)
+		}
+		time.Sleep(time.Duration(sum) * time.Second)
+		sum += 2
+	}
+
+	processServicePings(servicePings)
+
+	log.Debug("Done fetching active services.")
+}
+
+func main() {
+	var err error
+	var sum int
+	var zoneId string
+
+	if len(os.Args) != 3 {
+		err = errors.New(fmt.Sprintf("Usage: %s [domain] [Redis host]\n", os.Args[0]))
+		logErrorAndFail(err)
+	}
+
+	DNSName = os.Args[1]
+	configuration.RedisHost = os.Args[2]
+
+	sum = 1
+	for {
+		// We try to get the Hosted Zone Id using exponential backoff
+		zoneId, err = getDNSHostedZoneId()
+		if err == nil {
+			break
+		}
+		if sum > 8 {
+			logErrorAndFail(err)
+		}
+		time.Sleep(time.Duration(sum) * time.Second)
+		sum += 2
+	}
+	configuration.HostedZoneId = zoneId
+
+	sum = 1
+	for {
+		// We try to get the Redis connection using exponential backoff
+		redisClient = redis.NewClient(&redis.Options{
+			Addr:       configuration.RedisHost + ":6379",
+			Password:   "", // none
+			DB:         0,  // default DB
+			MaxRetries: 3,
+		})
+		// Check the connection.
+		err := redisClient.Ping().Err()
+
+		if err == nil {
+			break
+		}
+		if sum > 8 {
+			logErrorAndFail(err)
+		}
+		time.Sleep(time.Duration(sum) * time.Second)
+		sum += 2
+	}
+
+	// check regularly, specified by checkInterval
+	ticker := time.NewTicker(checkInterval)
```

Comment:
This part is much easier to swap out than redecs_agent which I just deployed, so I'll add a TODO here for this enhancement.